api/pipeline_spec.proto

```diff
@@ -0,0 +1,420 @@
+  message ImporterSpec {
+    // The URI of the artifact.
+    string artifact_uri = 1;
+
+    // The type of the artifact.
+    ArtifactTypeSchema type_schema = 2;
+
+    // The properties of the artifact.
+    map<string, ValueOrRuntimeParameter> properties = 3;
+
+    // The custom properties of the artifact.
+    map<string, ValueOrRuntimeParameter> custom_properties = 4;
+
+    // Whether or not import an artifact regardless it has been imported before.
+    bool reimport = 5;
+  }
+
+  // ResolverSpec is subject to change. Currently we only use enum to
+  // represent two of the currently available policies. We plan to introduce a
+  // flexible config to enable more sophisticated policies in the future.
+  // TODO(b/152230663): Support more flexibility for resolution logic.
+  message ResolverSpec {
+    enum ResolverPolicy {
+      RESOLVER_POLICY_UNSPECIFIED = 0;
+      // Within the models associated with the same pipeline context, always
+      // select the model with largest `last_update_time_since_epoch`.
+      LATEST_MODEL = 1;
+      // Within the models associated with the same pipeline context,
+      // select the latest Model that is blessed (represented by having a
+      // corresponding ModelBlessing artifact. The result will be consumed by
+      // components that need the latest model that is blessed such as
+      // Evaluator.
+      LATEST_BLESSED_MODEL = 2;
+    }
+    // TODO(b/152230663): Make this one of special resolver policies or generic
+    // ResolverConfig.
+    ResolverPolicy resolver_policy = 1;
+  }
+
+  // The specification of the executor.
+  message ExecutorSpec {
+    oneof spec {
+      // Starts a container.
+      PipelineContainerSpec container = 1;
+      // Import an artifact.
+      ImporterSpec importer = 2;
+      // Resolves an existing artifact.
+      ResolverSpec resolver = 3;
+    }
+  }
+  // Map from executor label to executor spec.
+  map<string, ExecutorSpec> executors = 1;
+}
+
+// Value is the value of the field.
+message Value {
+  oneof value {
+    // An integer value
+    int64 int_value = 1;
+    // A double value
+    double double_value = 2;
+    // A string value
+    string string_value = 3;
+  }
+}
+
+// The definition of a runtime artifact.
+message RuntimeArtifact {
+  // The name of an artifact.
+  string name = 1;
+
+  // The type of the artifact.
+  ArtifactTypeSchema type = 2;
+
+  // The URI of the artifact.
+  string uri = 3;
+
+  // The properties of the artifact.
+  map<string, Value> properties = 4;
+
+  // The custom properties of the artifact.
+  map<string, Value> custom_properties = 5;
+}
+
+// The input of an executor, which includes all the data that
+// can be passed into the executor spec by a string based placeholder.
+//
+// The string based placeholder uses a JSON path to reference to the data
+// in the [ExecutionInput]().
+//
+// `{{$}}`: prints the full [ExecutorInput]() as a JSON string.

```

Comment:
Is there a strict specification on the format of this placeholder?

Reply:
This is not a flexible templating language like Jinja. The list includes all the supported placeholders.